lib/unison-util-relation/src/Unison/Util/Relation.hs

```diff
@@ -500,44 +502,48 @@ compactSet = S.fold (S.union . fromMaybe S.empty) S.empty
 -- | Domain restriction for a relation. Modeled on z.
 (<|), restrictDom :: (Ord a, Ord b) => Set a -> Relation a b -> Relation a b
 restrictDom = (<|)
-s <| r =
-  fromList $
-    concatMap (\(x, y) -> zip (repeat x) (S.toList y)) (M.toList domain')
+s <| r = go s (domain r)
   where
-    domain' = M.unions . List.map filtrar . S.toList $ s
-    filtrar x = M.filterWithKey (\k _ -> k == x) dr
-    dr = domain r -- just to memoize the value
+    go _ Map.Tip = mempty
+    go s _ | Set.null s = mempty
+    go s (Map.Bin _ amid bs l r) = here <> go sl l <> go sr r
+      where
+        (sl, sr) = Set.split amid s
```

Comment:
You can use [`Set.splitMember`](https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Set.html#v:splitMember) to skip the `Set.member amid s` below and do it all in a single step 😄 .

Reply:
Oh, nice! Did not know